File: backend/src/main/java/xyz/engsmyre/moviedescriptiongame/db/GameSessionMovies.java
package xyz.engsmyre.moviedescriptiongame.db;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import org.springframework.data.annotation.Id;
import org.springframework.data.redis.core.RedisHash;
import org.springframework.data.redis.core.index.Indexed;

@RedisHash("Session")
public class GameSessionMovies implements Serializable {
    @Id
    @Indexed
    private UUID sessionId;
    private List<MovieStorage> movies;

    GameSessionMovies() {
        if (this.movies == null) {
            this.movies = new ArrayList<>();
        }
    }

    public GameSessionMovies(UUID sessionId) {
        this(new ArrayList<>(), sessionId);
    }

    public GameSessionMovies(List<MovieStorage> movies, UUID sessionId) {
        this.movies = movies;
        this.sessionId = sessionId;
    }

    public GameSessionMovies(GameSessionMovies oldSession, MovieStorage newMovie) {
        List<MovieStorage> newMovieList = new ArrayList<>(oldSession.getMovies());
        newMovieList.add(newMovie);
        this.movies = newMovieList;
        this.sessionId = oldSession.getSessionId();
    }

    public List<MovieStorage> getMovies() {
        return movies;
    }

    public UUID getSessionId() {
        return sessionId;
    }

    @Override
    public String toString() {
        return "GameSessionMovies{" +
                "sessionId=" + sessionId +
                ", movies=" + movies +
                '}';
    }
}

File: backend/src/main/kotlin/xyz/engsmyre/moviedescriptiongame/db/GameSessionMovies.kt
package xyz.engsmyre.moviedescriptiongame.db

import org.springframework.data.annotation.Id
import org.springframework.data.redis.core.RedisHash
import org.springframework.data.redis.core.index.Indexed
import java.io.Serializable
import java.util.*

@RedisHash("Session")
class GameSessionMovies : Serializable {
    @Id
    @Indexed
    var sessionId: UUID? = null
        private set
    var movies: List<MovieStorage>? = null
        private set

    internal constructor() {
        if (movies == null) {
            movies = ArrayList()
        }
    }

    constructor(sessionId: UUID?) : this(ArrayList<MovieStorage>(), sessionId) {}
    constructor(movies: List<MovieStorage>?, sessionId: UUID?) {
        this.movies = movies
        this.sessionId = sessionId
    }

    constructor(oldSession: GameSessionMovies, newMovie: MovieStorage) {
        val newMovieList: MutableList<MovieStorage> = ArrayList(oldSession.movies)
        newMovieList.add(newMovie)
        movies = newMovieList
        sessionId = oldSession.sessionId
    }

    override fun toString(): String {
        return "GameSessionMovies(sessionId=$sessionId, movies=$movies)"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as GameSessionMovies

        if (sessionId != other.sessionId) return false
        if (movies != other.movies) return false

        return true
    }

    override fun hashCode(): Int {
        var result = sessionId?.hashCode() ?: 0
        result = 31 * result + (movies?.hashCode() ?: 0)
        return result
    }


}